# Tarea 1 Bases de Datos 2023-1 Grupo 19



## Integrantes:
* Gustav Jacobsson - 202290241-8
* Kaled Hernandez - 202173025-7

## Requisitos para Correr el Notebook

Para poder utilizar este Notebook se debe tener instalado:
- **Python 3**
- **Jupyter**
- La librería **ipython-sql** (Poder ejecutar sql en Jupyter Notebook)
- PostgreSQL y su conector desde Python(**psycopg2**)

**IMPORTANTE:** Este Jupyter Notebook se conectará a su servicio de Base de Datos PostgreSQL de manera **local**, por lo que **no funcionará correctamente en el entorno Google Collab.**

Para instalar las dependencias, ejecute (**recuerde actualizar pip3**):

In [1]:
! pip3 install ipython-sql
! pip3 install psycopg2

#Use to generate names
! pip3 install names


[notice] A new release of pip available: 22.2.1 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip available: 22.2.1 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip available: 22.2.1 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


## Importar dependencias

In [2]:
from datetime import datetime
import random as r
import names

#Librerías para PostgreSQL
import psycopg2

Para conectarse a la base de datos, completen los datos relevantes en la variable DATABASE_URL.

* Cambie `user` por el usuario que utilizó en Postgres para acceder a la Base de Datos. Por defecto `postgres`.
* Cambie `password` por la contraseña de dicho usuario. **Recuerde borrarla antes de entregar su tarea.**
* Cambie `port` por el puerto en que se ejecuta su servicio de Base de Datos. Por defecto `5432`.
* Cambie `database` por el nombre de la BD (Schema) que creó en Postgres.

Ejemplo: `postgresql://postgres:pass_segura@localhost:5432/tarea-bd`

In [3]:
%reload_ext sql
%env DATABASE_URL=postgresql://postgres:papelucho1@localhost:5432/Tarea1DB

env: DATABASE_URL=postgresql://postgres:password@localhost:5432/Tarea1DB


Para conectarse de forma declarativa a la Base de Datos, utilice el siguiente código:

In [5]:
#Recuerde cambiar los valores de USER, DBNAME y PASSWORD por los mismos de la celda superior.
connection = psycopg2.connect(
   host="localhost",
   user="postgres",
   dbname='Tarea1DB',
   password="papelucho1"
)

print(connection)
cursor = connection.cursor()

<connection object at 0x0000019FA6857340; dsn: 'user=postgres password=xxx dbname=Tarea1DB host=localhost', closed: 0>


## Crear tablas (36 pts.)

Aquí inserta el código SQL que creara las tablas de la base de datos:

### Tabla "facciones" (6 pts)

In [13]:
%%sql CREATE TABLE facciones(
	id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
	nombre VARCHAR(45) NOT NULL,
	descripcion VARCHAR(45) NOT NULL
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

### Tabla "trabajos" (6 pts)

In [14]:
%%sql CREATE TABLE trabajos(
	id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
	nombre VARCHAR(45) NOT NULL,
	descripcion VARCHAR(45) NOT NULL,
    sueldo INTEGER NOT NULL	
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

### Tabla "personas" (6 pts)

In [16]:
%%sql CREATE TABLE personas(
	id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
	nombre VARCHAR(45) NOT NULL,
    apellidos VARCHAR(45) NOT NULL,
    fecha_nacimiento DATE NOT NULL,
	infectado BOOL NOT NULL,
    conyugue INTEGER,
    FOREIGN KEY (conyugue) REFERENCES personas (id)
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

### Tabla "cuentas_bancarias" (6 pts)

In [17]:
%%sql CREATE TABLE cuentas_bancarias(
	id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
	tipo_de_cuenta VARCHAR(45) NOT NULL,
	banco VARCHAR(45) NOT NULL,
    saldo INTEGER NOT NULL,
    id_persona INTEGER NOT NULL,
    FOREIGN KEY (id_persona) REFERENCES personas (id)
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

### Tabla "persona_pertenece_faccion" (6 pts)

In [18]:
%%sql CREATE TABLE persona_pertenece_faccion(
	id_persona INTEGER,
    id_faccion INTEGER,
    FOREIGN KEY (id_persona) REFERENCES personas(id),
    FOREIGN KEY (id_faccion) REFERENCES facciones(id),
	PRIMARY KEY(id_persona,id_faccion)
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

### Tabla "persona_tiene_trabajo" (6 pts)

In [25]:
%%sql CREATE TABLE persona_tiene_trabajo(
	id_persona INTEGER,
    id_trabajo INTEGER,
    estado BOOL NOT NULL,
    ultima_vez_realizado TIMESTAMP NOT NULL,
    FOREIGN KEY (id_persona) REFERENCES personas(id)
);


 * postgresql://postgres:***@localhost:5432/Tarea1DB
Done.


[]

## Código Generador de Datos

Ejecute este código para cargar datos a las tablas creadas.

PD: Si el codigo tira error por "violar la unicidad de las PK" es por una coincidencia de numeros random, solamente volver a ejecutar el codigo

In [26]:
#Recuerde haber ejecutado todas las celdas anteriores!

# Generador de timestamp aleatorio.
def generate_timestamp_nacimiento():
    year = r.randint(1970,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

def generate_timestamp_trabajo():
    year = r.randint(2019,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

print("Borrando Datos Antiguos...")

# Limpia la base de datos.
connection.commit() # En caso de algun error
query = 'TRUNCATE TABLE "facciones", "trabajos", "personas", "cuentas_bancarias", "persona_pertenece_faccion", "persona_tiene_trabajo" RESTART IDENTITY'
cursor.execute(query)
connection.commit()

print("Generando Datos...")

facciones = [("FEDRA", "La Agencia Federal de Respuesta a Desastres"), ("Luciérnagas", "Grupo de milicias revolucionario"), ("Carroñeros", "Sin identificación"), ("Pythonia", "Grupo defensor de la libertad"), ("MaCdonia", "Los exiliados de la carceles"), ("JavaScriptia", "Grupo defensor de la paz")]
trabajos = [("Limpieza", "Encargado de limpiar las calles"),
("Limpiador de cloacas","Encargado de limpiar las cloacas"),
("Militar","Encargado de proteger la ciudad"),
("Explorador","Encargado de explorar la ciudad"),
("Médico","encargado de curar a los enfermos"),
("Profesor","Encargado de enseñar a los niños"),
("Cremador","Encargado de quemar los cadáveres"),
("Científico","Encargado de investigar"),
("Cocinero","Encargado de cocinar"),
("Carpintero","Encargado de construir"),
("Cerrajero","Encargado de reparar las cerraduras"),
("Cartero","Encargado de repartir los correos"),
("Cajero","Encargado de cobrar los impuestos")]
tipos_de_cuenta_bancaria = ["Cuenta Vista", "Cuenta Ahorro", "Cuenta Corriente", "Cuenta RUT", "Cuenta de Inversion", "Cuenta PrePago", "Cuenta de Credito"]
bancos = ["FedraBank S.A", "Banco Pedrito Pascal SpA", "Banco Pythonia S.A", "Banco de la Nación S.A", "Banco Santander SpA", "Banco BCI SpA", "Banco Estado SpA", "Banco Consorcio S.A", "Banco BBVA", "Banco Scotiabank"]

# Variables.
n_facciones = len(facciones)
n_trabajos = len(trabajos)
n_personas = 1000
n_cuentas_bancarias = 120
n_persona_pertenece_faccion = 300
n_persona_tiene_trabajo = 300
n_casados = 100

# Genera personas.
personas = []
for _ in range(n_personas):
    nombre = names.get_first_name()
    apellidos = names.get_last_name()+" "+names.get_last_name()
    fecha_nacimiento = generate_timestamp_nacimiento()
    infectado = r.choice([True, False])
    
    personas.append((nombre, apellidos, fecha_nacimiento, infectado, None))


# Genera cuentas bancarias.
cuentas_bancarias = []
for _ in range(n_cuentas_bancarias):
    tipo_de_cuenta = r.choice(tipos_de_cuenta_bancaria)
    banco = r.choice(bancos)
    saldo = r.randint(0, 10000)
    id_persona = r.choice(range(1, n_personas+1))
    
    cuentas_bancarias.append((tipo_de_cuenta, banco, saldo, id_persona))
    
    
# Generate persona_pertenece_faccion.
persona_pertenece_faccion = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_facciones+1))) for _ in range(1, n_persona_pertenece_faccion+1)])

# Generate persona_tiene_trabajo.
persona_tiene_trabajo_prev = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_trabajos+1))) for _ in range(1, n_persona_tiene_trabajo+1)])
persona_tiene_trabajo = [(id_persona, id_trabajo, r.choice([True, False]), generate_timestamp_trabajo()) for id_persona, id_trabajo in persona_tiene_trabajo_prev]
    
# SQL Statements.
inserts = [
    'INSERT INTO "facciones" ("nombre", "descripcion") VALUES (%s, %s);',
    'INSERT INTO "trabajos" ("nombre", "descripcion", "sueldo") VALUES (%s, %s, %s);',
    'INSERT INTO "personas" ("nombre", "apellidos", "fecha_nacimiento", "infectado", "conyugue") VALUES (%s, %s, %s, %s, %s);',
    'INSERT INTO "cuentas_bancarias" ("tipo_de_cuenta", "banco", "saldo", "id_persona") VALUES (%s, %s, %s, %s);',
    'INSERT INTO "persona_pertenece_faccion" ("id_persona", "id_faccion") VALUES (%s, %s);',
    'INSERT INTO "persona_tiene_trabajo" ("id_persona", "id_trabajo", "estado", "ultima_vez_realizado") VALUES (%s, %s, %s, %s);'
]

for faccion in facciones:
    cursor.execute(inserts[0], faccion)

for trabajo in trabajos:
    t = (trabajo[0], trabajo[1], r.randint(100, 5000))
    cursor.execute(inserts[1], t)

# Crear una lista con personas casadas y otra con personas solteras. donde se cambie la ide de la persona que se casó.

for persona in personas:
    cursor.execute(inserts[2], persona)
    
for cb in cuentas_bancarias:
    cursor.execute(inserts[3], cb)

for ppf in persona_pertenece_faccion:
    cursor.execute(inserts[4], ppf)
    
for ptt in persona_tiene_trabajo:
    cursor.execute(inserts[5], ptt)
    
connection.commit()


# Generar Matrimonios
n_repetidos = []

for i in range(n_casados):
    random = r.randint(1, n_personas)
    if random in n_repetidos:
        continue
    else:
        random2 = r.randint(1, n_personas)
        if random2 in n_repetidos:
            continue
        else:
            n_repetidos.append(random)
            n_repetidos.append(random2)
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random2, random))
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random, random2))
            connection.commit()

print("Todo Listo!")

Borrando Datos Antiguos...
Generando Datos...
Todo Listo!


## Consultas (64 pts.)

Escriba aquí las consultas pedidas. Recuerde agregar una descripción _breve_ de cómo funciona la solución. **Escriba las consultas de forma ordenada.**

**NOTA:** Cuando en una consulta diga "X cosa" significa que ustedes deben de elegir ese X por lo que para cualquier otro X debe de funcionar de igual forma.

### Consulta 1 (6 pts)
**ID, nombre y primer apellido de los habitantes de FEDRA**

Ejemplo:

|ID |  Nombre      |  Apellido  |
|---|-----------|-----------|
| 12 |Joel | Miller |
|21 | Ellie | Williams |
| ... | ... | ... |

**Solución:** Se crea una tabla mas grande con personas, persona_pertence_faccion y facciones para poder acceder a todos los datos, se usa split_part para mostrar solo el primer apellido de las personas


In [36]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    personas.id, personas.nombre, SPLIT_PART(personas.apellidos, ' ' ,1) AS apellidos
FROM 
    personas JOIN persona_pertenece_faccion ON personas.id = persona_pertenece_faccion.id_persona 
    JOIN facciones ON persona_pertenece_faccion.id_faccion = facciones.id
WHERE
    persona_pertenece_faccion.id_faccion = 1;

 * postgresql://postgres:***@localhost:5432/Tarea1DB
57 rows affected.


[(21, 'Margarita', 'Thompson'),
 (33, 'Marian', 'Guardado'),
 (59, 'Brenda', 'Mehtani'),
 (32, 'Aurora', 'Lynch'),
 (132, 'John', 'Fitzgerald'),
 (137, 'Marie', 'Mcallen'),
 (172, 'Sarah', 'Rodriguez'),
 (220, 'Travis', 'Rock'),
 (222, 'Michele', 'Sitter'),
 (231, 'Melissa', 'Perry'),
 (185, 'Cynthia', 'Vanbuskirk'),
 (264, 'Salena', 'Ratterman'),
 (312, 'Peter', 'Guerra'),
 (329, 'Maria', 'Brobst'),
 (321, 'Lisa', 'Olson'),
 (379, 'Monty', 'Debois'),
 (394, 'Wm', 'Butts'),
 (406, 'William', 'Bowen'),
 (437, 'Peter', 'Butler'),
 (439, 'Brad', 'Snowden'),
 (485, 'Robert', 'Chapman'),
 (423, 'Joan', 'Harris'),
 (407, 'Michael', 'Perry'),
 (517, 'Colleen', 'Graves'),
 (524, 'Eric', 'Curtis'),
 (533, 'Frank', 'Olsen'),
 (544, 'Keith', 'Zimmerman'),
 (581, 'Latrice', 'Mcelroy'),
 (566, 'James', 'Reed'),
 (637, 'Michelle', 'Lombardo'),
 (645, 'Chelsea', 'Creekmore'),
 (652, 'Stacey', 'Moseley'),
 (663, 'Daphne', 'Jones'),
 (687, 'Kathryn', 'Gray'),
 (707, 'Donna', 'Morvillo'),
 (710, 'Ronda', 'Sartoris'),
 (715, 'Joel', 'Henley'),
 (721, 'Helena', 'Liberti'),
 (729, 'John', 'Wheaton'),
 (746, 'Johnny', 'White'),
 (759, 'Michael', 'Simmons'),
 (832, 'Paul', 'Feuerborn'),
 (864, 'Sandra', 'Patten'),
 (869, 'Eugene', 'Harold'),
 (872, 'Robert', 'Glass'),
 (824, 'Christine', 'Thomas'),
 (889, 'James', 'Gutierrez'),
 (890, 'Marjorie', 'Tillman'),
 (908, 'John', 'Kilmer'),
 (909, 'Wilbert', 'Weaver'),
 (918, 'Kathy', 'Jordan'),
 (931, 'Tonya', 'Hoskins'),
 (986, 'Marlon', 'Hale'),
 (887, 'Nicholas', 'Plaza'),
 (927, 'Herbert', 'Johnson'),
 (894, 'Christine', 'Chambers'),
 (424, 'Jeannette', 'Cornish')]

### Consulta 2 (6 pts)
**Cantidad de personas nacidas después del inicio de la pandemia.**(inicio de la pandemia 01-12-2019)

Ejemplo:


| Cantidad     | 
|--------------|
| 292 | 

**Solución:**Se cuenta a las personas con fecha de nacimiento con COUNT y se filtran con la seccion WHERE

In [37]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    COUNT(personas.fecha_nacimiento) AS Cantidad
FROM 
    personas
WHERE
    personas.fecha_nacimiento > '2019-12-01'

 * postgresql://postgres:***@localhost:5432/Tarea1DB
1 rows affected.


[(44,)]

### Consulta 3 (6 pts)
**ID, nombre, apellidos y saldo total de todas las personas**( Limitar a 20 filas para visualizar)

Ejemplo:

| ID    | Nombre | Apellido | Dinero | 
|----------------|--------|-|-|
| 13 | Tommy | Miller | 12501 |
| ... | ... | ... | ... |

**Solución:** Se accede a la informacion directamente de la tabla creada del JOIN entre personas y cuentas_bancarias

In [10]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

SELECT
    personas.id, personas.nombre, SPLIT_PART(personas.apellidos, ' ' ,1) AS apellido, cuentas_bancarias.saldo
FROM 
    personas JOIN cuentas_bancarias ON personas.id = id_persona
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/Tarea1DB
20 rows affected.


[(8, 'Vernon', 'Cliff', 2567),
 (11, 'Patrick', 'Hill', 4014),
 (28, 'Ashley', 'Hancock', 8733),
 (53, 'Melvin', 'Barbee', 6762),
 (56, 'Bruce', 'Reed', 8652),
 (62, 'Stephanie', 'Evans', 2662),
 (63, 'David', 'Griffin', 4646),
 (64, 'Lana', 'Kelly', 643),
 (64, 'Lana', 'Kelly', 4146),
 (80, 'Holly', 'Stewart', 8327),
 (88, 'Rebecca', 'Helfer', 2112),
 (93, 'Harry', 'Denmon', 6270),
 (94, 'Andrew', 'Grimes', 5405),
 (97, 'John', 'Bond', 7836),
 (104, 'Aaron', 'Ward', 9804),
 (105, 'Shannon', 'Schumacher', 5393),
 (108, 'Guadalupe', 'Hanford', 171),
 (111, 'John', 'Turpen', 1830),
 (114, 'Gerald', 'Dorn', 3891),
 (4, 'Lily', 'Bowen', 6779)]

### Consulta 4 (6 pts)
**Nombre de la Facción y la cantidad de trabajadores activos hasta fecha X.**

Ejemplo:

| Facción     | Cantidad de trabajadores |
|----------------|--------|
| FEDRA | 1273 |
| Luciérnagas | 121 |
| Carroñeros  | 72 |
| ...  | ... |

**Solución:** Se crea una tabla entre personas, persona_pertenece_faccion, facciones y persona_tiene_trabajo y se cuentan la cantidad de trabajadores, se filtran los datos en la seccion WHERE y se agrupan en la seccion facciones

In [11]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. 2021-09-11 DUDAS*/

SELECT
    facciones.nombre AS Facción, COUNT(*) AS "Cantidad de trabajadores"
FROM
    personas JOIN persona_pertenece_faccion ON personas.id = persona_pertenece_faccion.id_persona
    JOIN facciones ON persona_pertenece_faccion.id_faccion = facciones.id
    JOIN persona_tiene_trabajo ON personas.id = persona_tiene_trabajo.id_persona
WHERE
    DATE(persona_tiene_trabajo.ultima_vez_realizado) < '2021-09-11' AND persona_tiene_trabajo.estado = true
GROUP BY
    facciones.nombre
ORDER BY
    COUNT(*) DESC;


 * postgresql://postgres:***@localhost:5432/Tarea1DB
6 rows affected.


[('FEDRA', 8),
 ('MaCdonia', 7),
 ('Carroñeros', 5),
 ('JavaScriptia', 5),
 ('Luciérnagas', 4),
 ('Pythonia', 3)]

### Consulta 5 (6 pts)
**Nombre de la Facción y su cantidad de personas infectadas.**

Ejemplo:

|  Facción  | Personas infectadas | 
|--------|-------|
| FEDRA | 251 |
| Luciérnagas | 17 |
| Carroñeros  | 12 |
| ...  | ... |

**Solución:** JOIN entre personas, persona_pertenece_faccion y facciones, se accede a la cantidad de personas infectadas con COUNT y se filtran con perosnas.infectados = true, se agrupan en facciones.nombre

In [12]:
%%sql

SELECT
    facciones.nombre AS Facción, COUNT(*) AS "Personas infectadas"
FROM
    personas JOIN persona_pertenece_faccion ON personas.id = persona_pertenece_faccion.id_persona
    JOIN facciones ON persona_pertenece_faccion.id_faccion = facciones.id
WHERE
    personas.infectado = true
GROUP BY
    facciones.nombre
ORDER BY
    COUNT(*) DESC;

 * postgresql://postgres:***@localhost:5432/Tarea1DB
6 rows affected.


[('FEDRA', 32),
 ('JavaScriptia', 23),
 ('Carroñeros', 22),
 ('MaCdonia', 21),
 ('Luciérnagas', 17),
 ('Pythonia', 17)]

### Consulta 6 (6 pts)
**Nombre de la Facción y su trabajo más realizado.**

Ejemplo:

|  Facción  | Trabajo | 
|--------|-------|
| FEDRA | Limpiador de cloacas |
| Luciérnagas | Médico |
| Carroñeros  | Explorador |
| ...  | ... |

**Solución:** _Firstly the faccion.nombre is selected. A inner query is then used to select 1 and only 1 "trabajo", that being the one occurring most in the list (ORDER BY COUNT()). An inner query is used as when both the faction name and trabajo was selected at the same time it returned various "trabajos" for each faction._

In [5]:
%%sql


SELECT
   facciones.nombre, 
   (SELECT trabajos.nombre
   FROM persona_pertenece_faccion
   INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id
   INNER JOIN persona_tiene_trabajo ON personas.id = persona_tiene_trabajo.id_persona
   INNER JOIN trabajos ON persona_tiene_trabajo.id_trabajo = trabajos.id
   WHERE persona_pertenece_faccion.id_faccion = facciones.id
   GROUP BY trabajos.nombre
   ORDER BY COUNT(*) DESC
   LIMIT 1)
FROM facciones;




 * postgresql://postgres:***@localhost:5432/Tarea1DB
6 rows affected.


[('FEDRA', 'Explorador'),
 ('Luciérnagas', 'Explorador'),
 ('Carroñeros', 'Explorador'),
 ('Pythonia', 'Explorador'),
 ('MaCdonia', 'Explorador'),
 ('JavaScriptia', 'Explorador')]

### Consulta 7 (8 pts)
**Nombre de la Facción y su Saldo total** (Es decir, la suma total de los saldos de cada persona perteneciente a la facción.)

Ejemplo:

|Facción | Saldo |
|--------|-|
| FEDRA | 523000 |
| Luciérnagas | 130023 |
| Carroñeros  | 17432 |
| ...  | ...|

**Solución:** Each faction is INNER JOINED with "perosna_pertenece_faccion" and then with "personas" and lastly with "cuenta_bancaria". These are then grouped by faction, and "SUM(cuenta_bancaria.sueldo) is displayed._

In [4]:
%%sql


SELECT
    facciones.nombre, SUM(cuentas_bancarias.saldo) AS saldo_total
FROM facciones
    INNER JOIN persona_pertenece_faccion ON facciones.id = persona_pertenece_faccion.id_faccion
    INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id
    INNER JOIN cuentas_bancarias ON personas.id = cuentas_bancarias.id_persona
    GROUP BY facciones.nombre;

6 rows affected.


[('JavaScriptia', 26445),
 ('Pythonia', 63191),
 ('MaCdonia', 31281),
 ('FEDRA', 14427),
 ('Luciérnagas', 37821),
 ('Carroñeros', 10482)]

### Consulta 8 (6 pts)
**Nombre del Banco, cantidad de usuarios pertenecientes al banco y saldo total del banco.**

###### Es decir, todos los bancos que ocupan los habitantes de cada faccion

Ejemplo:

|Banco | Cantidad|Saldo|
|----|---|---|
|FedraBank S.A |321|20000|
|Banco Pedrito Pascal SpA  |21|200| 
| Banco De Chile|213|56000|
| ... | ... |

**Solución:** _The bank name (cuentas_bancarias.banco), the number of people using that list (COUNT(*), which basically is how many time each bank-name apperas in the table), and the SUM (which is the total of "saldo" for each bank name) is selected. They are then grouped by cuentas_bancarias.banco which makes one "number of users" and one "saldo" for each bank.

In [6]:
%%sql

SELECT
    cuentas_bancarias.banco, COUNT(*) AS "Cantidad", SUM(cuentas_bancarias.saldo) AS "Saldo"
FROM cuentas_bancarias
GROUP BY cuentas_bancarias.banco

 * postgresql://postgres:***@localhost:5432/Tarea1DB
10 rows affected.


[('Banco Consorcio S.A', 14, 69250),
 ('Banco de la Nación S.A', 8, 40277),
 ('Banco Santander SpA', 9, 64600),
 ('Banco Pythonia S.A', 20, 113377),
 ('Banco BBVA', 9, 45384),
 ('Banco Estado SpA', 12, 72308),
 ('Banco BCI SpA', 13, 66261),
 ('FedraBank S.A', 14, 79038),
 ('Banco Pedrito Pascal SpA', 11, 53847),
 ('Banco Scotiabank', 10, 50622)]

### Consulta 9 (8 pts)
**Nombre de la Facción y el Banco más utilizado por sus habitantes**(Si hay empate, que se muestre solo uno.)

Ejemplo:
   
|Facción | Banco |
| - | - |
| FEDRA | FedraBank S.A |
| Luciérnagas | Banco Pedrito Pascal SpA |
| ... | ... |

**Solución:** _Similar to the solution in consulta 6. An inner join is used to get 1 and only 1 bank, which is the most popular bank within a faction. Since only faction.name is selected in the outer query then each faction is displayed together with the most popular bank for that faction._

In [7]:
%%sql
SELECT 
    facciones.nombre,
    (SELECT cuentas_bancarias.banco
     FROM persona_pertenece_faccion
     INNER JOIN personas ON persona_pertenece_faccion.id_persona = personas.id
     INNER JOIN cuentas_bancarias ON personas.id = cuentas_bancarias.id_persona
     WHERE persona_pertenece_faccion.id_faccion = facciones.id
     GROUP BY cuentas_bancarias.banco
     ORDER BY COUNT(*) DESC
     LIMIT 1) as banco_mas_popular
FROM facciones;

 * postgresql://postgres:***@localhost:5432/Tarea1DB
26 rows affected.


[('Pythonia', 'FedraBank S.A'),
 ('FEDRA', 'Banco Estado SpA'),
 ('Pythonia', 'Banco Consorcio S.A'),
 ('JavaScriptia', 'Banco Pythonia S.A'),
 ('Pythonia', 'Banco BCI SpA'),
 ('MaCdonia', 'Banco BBVA'),
 ('JavaScriptia', 'Banco BBVA'),
 ('Luciérnagas', 'Banco Scotiabank'),
 ('MaCdonia', 'Banco Consorcio S.A'),
 ('Luciérnagas', 'Banco Pythonia S.A'),
 ('Carroñeros', 'Banco Estado SpA'),
 ('FEDRA', 'Banco Pedrito Pascal SpA'),
 ('MaCdonia', 'Banco Pedrito Pascal SpA'),
 ('Luciérnagas', 'FedraBank S.A'),
 ('MaCdonia', 'Banco Estado SpA'),
 ('Pythonia', 'Banco Scotiabank'),
 ('Luciérnagas', 'Banco Santander SpA'),
 ('Luciérnagas', 'Banco BBVA'),
 ('Carroñeros', 'Banco Pythonia S.A'),
 ('MaCdonia', 'Banco Scotiabank'),
 ('JavaScriptia', 'Banco Scotiabank'),
 ('Pythonia', 'Banco Santander SpA'),
 ('MaCdonia', 'FedraBank S.A'),
 ('Luciérnagas', 'Banco Consorcio S.A'),
 ('FEDRA', 'Banco de la Nación S.A'),
 ('Pythonia', 'Banco Pythonia S.A')]

### Consulta 10 (6 pts)
**ID, nombre y segundo apellido de las personas traidoras**(Es decir, si la persona pertenece a dos o más facciones.)

Ejemplo:
   
|ID | Nombre | Apellido | 
|-|-------|-|
| 1 | Pedro | Pascal |
| 2 | Bella | Ramsey |
| ... | ... | ... | 

**Solución:** _Each persons ID, Name and Apellido is selected. They are then inner joined with the "persona_pertenece_faccion". Using the HAVING COUNT(persona_pertenece_facicion-id_faccion) > 1 we can see if any person has more than 1 faction connected to them in the inner join table.._

In [8]:
%%sql

SELECT p.id as "ID", p.nombre, SPLIT_PART(p.apellidos, ' ', 1) AS apellido
FROM personas p
INNER JOIN persona_pertenece_faccion ON p.id = persona_pertenece_faccion.id_persona
GROUP BY p.id, p.nombre, SPLIT_PART(p.apellidos, ' ', 1)
HAVING COUNT(persona_pertenece_faccion.id_faccion) > 1;

 * postgresql://postgres:***@localhost:5432/Tarea1DB
24 rows affected.


[(21, 'Margarita', 'Thompson'),
 (759, 'Michael', 'Simmons'),
 (707, 'Donna', 'Morvillo'),
 (742, 'Ruth', 'Nordstrom'),
 (454, 'William', 'Spore'),
 (116, 'Erica', 'Pershall'),
 (524, 'Eric', 'Curtis'),
 (85, 'Maria', 'Foglesong'),
 (570, 'Elizabeth', 'Vess'),
 (291, 'Julie', 'Pipkins'),
 (765, 'Diana', 'Long'),
 (349, 'John', 'Flink'),
 (908, 'John', 'Kilmer'),
 (245, 'Linda', 'Farris'),
 (782, 'Jose', 'Anzalone'),
 (99, 'Gilbert', 'Mcgwin'),
 (48, 'Patricia', 'Reid'),
 (209, 'Kimberly', 'Esquilin'),
 (379, 'Monty', 'Debois'),
 (988, 'Margaret', 'Hesson'),
 (90, 'George', 'Noel'),
 (466, 'Allen', 'Jones'),
 (56, 'Bruce', 'Reed'),
 (941, 'Lillian', 'Greenhoward')]

### Consulta BONUS (6 pts.)

**Nombre y Apellidos de las personas casadas**(solo se muestra una única vez la pareja)

Ejemplo:
   
|Nombre persona 1 | Apellidos persona 1 | Nombre persona 2 | Apellidos persona 2 |
|-|-------|-|-|
| Felipe | Martinez | Raul | Cuello |
| Rodrigo | Munita | Juan | Mira |
| Nicole | Rojel | Benjamin | Campos |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [7]:
%%sql

SELECT
    personas.nombre, personas.apellido,
    (SELECT personas.nombre, personas.apellido 
     FROM personas p2
     WHERE p1.conyugue = p2.id
     GROUP BY p2.nombre, SPLIT_PART(p2.apellidos, ' ', 1).apellido
     LIMIT 1
    )    
FROM 
personas p1;


Traceback (most recent call last):
  File "c:\Users\Kaled\AppData\Local\Programs\Python\Python310\lib\site-packages\sqlalchemy\engine\base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
  File "c:\Users\Kaled\AppData\Local\Programs\Python\Python310\lib\site-packages\sqlalchemy\engine\base.py", line 3275, in raw_connection
    return self.pool.connect()
  File "c:\Users\Kaled\AppData\Local\Programs\Python\Python310\lib\site-packages\sqlalchemy\pool\base.py", line 455, in connect
    return _ConnectionFairy._checkout(self)
  File "c:\Users\Kaled\AppData\Local\Programs\Python\Python310\lib\site-packages\sqlalchemy\pool\base.py", line 1271, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "c:\Users\Kaled\AppData\Local\Programs\Python\Python310\lib\site-packages\sqlalchemy\pool\base.py", line 719, in checkout
    rec = pool._do_get()
  File "c:\Users\Kaled\AppData\Local\Programs\Python\Python310\lib\site-packages\sqlalchemy\pool\impl.py", lin